util/counter.go

```diff
@@ -0,0 +1,44 @@
+package util
+
+import (
+	"sync"
+)
+
+type AtomicCounter struct {
+	count     int
+	countLock *sync.RWMutex
+}
+
+type Counter interface {
+	GetCount() int
+	IncreaseCount()
+	DecreaseCount()
+}
+
+func NewAtomicCounter() Counter {
+	return &AtomicCounter{
+		countLock: &sync.RWMutex{},
+		count:     0,
+	}
+}
+
+func (ac *AtomicCounter) GetCount() int {
+	ac.countLock.RLock()
+	defer ac.countLock.RUnlock()
+
+	return ac.count
+}
+
+func (ac *AtomicCounter) IncreaseCount() {
+	ac.countLock.Lock()
+	defer ac.countLock.Unlock()
+
+	ac.count++
```

Comment:
@shuo-wu , @derekbit I don't know what is the benefit of using atomic here. Can you explain?

Reply:
Because the counter is only an integer, we can just simply use atomic operations to reduce the CPU cycles. The benefit should not be significant, so I mark it as nit.

Reply:
Agreed. We don't need to do the same thing again. (I forgot there are atomic things in golang natively..)

Reply:
Updated.